alicloud/resource_alicloud_mns_topic_subscription.go

```diff
@@ -0,0 +1,166 @@
+			"topic_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateStringLengthInRange(3, 256),
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateStringLengthInRange(3, 256),
+			},
+
+			"endpoint": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateEndpoint,
+			},
+
+			"filter_tag": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validateStringLengthInRange(0, 16),
+			},
+
+			"notify_strategy": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  string(ali_mns.BACKOFF_RETRY),
+				ValidateFunc: validateAllowedStringValue([]string{
+					string(ali_mns.BACKOFF_RETRY), string(ali_mns.EXPONENTIAL_DECAY_RETRY),
+				}),
+			},
+
+			"notify_content_format": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  string(ali_mns.SIMPLIFIED),
+				ForceNew: true,
+				ValidateFunc: validateAllowedStringValue([]string{
+					string(ali_mns.SIMPLIFIED), string(ali_mns.XML),
+				}),
+			},
+		},
+	}
+}
+
+func resourceAlicloudMNSSubscriptionCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	topicName := d.Get("topic_name").(string)
+	subscriptionManager, err := client.MnsSubscriptionManager(topicName)
+	if err != nil {
+		return fmt.Errorf("Creating mns subscription client  error: %#v", err)
+	}
+	name := d.Get("name").(string)
+	endpoint := d.Get("endpoint").(string)
+	notifyStrategyStr := d.Get("notify_strategy").(string)
+	notifyContentFormatStr := d.Get("notify_content_format").(string)
+	var filterTag string
+	if v, ok := d.GetOk("filter_tag"); ok {
+		filterTag = v.(string)
+	}
+	notifyStrategy := ConvertToNotifyStrategyType(notifyStrategyStr)
+	notifyContentFormat := ConvertToNotifyContentFormatType(notifyContentFormatStr)
+	subRequest := ali_mns.MessageSubsribeRequest{
+		Endpoint:            endpoint,
+		FilterTag:           filterTag,
+		NotifyStrategy:      notifyStrategy,
+		NotifyContentFormat: notifyContentFormat,
+	}
+	err = subscriptionManager.Subscribe(name, subRequest)
+	if err != nil {
+		return fmt.Errorf("Create Subscription got an error: %#v", err)
+	}
+	d.SetId(fmt.Sprintf("%s%s%s", topicName, COLON_SEPARATED, name))
+	return resourceAlicloudMNSSubscriptionRead(d, meta)
+}
+
+func resourceAlicloudMNSSubscriptionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	topicName, name := GetTopicNameAndSubscriptionName(d.Id())
+	subscriptionManager, err := client.MnsSubscriptionManager(topicName)
+	if err != nil {
+		return fmt.Errorf("Creating mns subscription client  error: %#v", err)
+	}
+	attr, err := subscriptionManager.GetSubscriptionAttributes(name)
+	if err != nil {
+		return fmt.Errorf("Get mns subscription attr   error: %#v", err)
```

Comment:
I think if error is TopciSubscriptionNotExist, line 112 should `d.setId("")` and `return nil`